docs/src/apireference.md

```diff
@@ -554,26 +554,44 @@ models. Nevertheless, you are still able to build incrementally an optimization
-2) Define a [`ModelLike`](@ref) `model_cache` that will store the incremental definition of the model.
-3) Cache `optimizer` with `model_cache` by calling `CachingOptimizer(model_cache, optimizer)`. This method returns an empty `AbstractOptimizer` `cached_optimizer`.
-4) Define your optimization model as usual by using `cached_optimizer` instead of `optimizer`. This will allocate the model directly to `model_cache` instead of `optimizer`. For instance, to add two variables in the model:
-```julia
-MOI.add_variables(cached_optimizer, 2)
-```
-5) Once the model properly defined, call [`Utilities.attach_optimizer`](@ref) to copy the optimization model all in once inside `optimizer`. This method uses the MathOptInterface method [`copy_to`](@ref) to copy information from `model_cache` to `optimizer`.
-
-Note that:
-- [`Utilities.drop_optimizer`](@ref) drops the underlying `optimizer` from `cached_optimizer`, without emptying it.
-- [`Utilities.reset_optimizer`](@ref) empties `optimizer` inside `cached_optimizer`, without droping it.
-
-Bridging correctly a solver with a `CachingOptimizer` requires to
-implement properly the [Allocate-Load API](@ref) inside the solver's
-MathOptInterface wrapper.
+The function [`Utilities.state`](@ref) allows to query the state
+of the optimizer cached inside a `CachingOptimizer`. The state
+could be:
+* `NO_OPTIMIZER`, if no optimizer is attached;
+* `EMPTY_OPTIMIZER`, if the attached optimizer is empty;
+* `ATTACHED_OPTIMIZER`, if the attached optimizer is synchronized with the
+  cached model defined in `CachingOptimizer`.
+
+Note that there exists methods to modify the attached optimizer:
+* [`Utilities.attach_optimizer`](@ref) attachs a new `optimizer`
+  to a `cached_optimizer` with state `EMPTY_OPTIMIZER`.
+  The state of `cached_optimizer` is set to `ATTACHED_OPTIMIZER` after the call.
+* [`Utilities.drop_optimizer`](@ref) drops the underlying `optimizer`
+  from `cached_optimizer`, without emptying it. The state of `cached_optimizer`
+  is set to `NO_OPTIMIZER` after the call.
+* [`Utilities.reset_optimizer`](@ref) empties `optimizer` inside
+  `cached_optimizer`, without droping it. The state of `cached_optimizer`
+  is set to `EMPTY_OPTIMIZER` after the call.
+
+The way to operate a `CachingOptimizer` depends whether the mode
+is set to `AUTOMATIC` or to `MANUAL`.
+* In `MANUAL` mode, the state of the `CachingOptimizer` changes only
+  if the methods [`Utilities.attach_optimizer`](@ref),
+  [`Utilities.reset_optimizer`](@ref) or [`Utilities.drop_optimizer`](@ref)
+  are being called.  Any unattended operation results in an error.
+* In `AUTOMATIC` mode, the state of the `CachingOptimizer` changes when
+  necessary. Any modification not supported by the solver (e.g. dropping
+  a constraint) results in a drop to the state `EMPTY_OPTIMIZER`.
+
+Wrapping correctly a solver with a `CachingOptimizer` requires to
+implement properly the [Allocate-Load API](@ref) inside the
```

Comment:
The solver should support `MOI.copy_to` but `MOI.copy_to` could
* either use the `add_constraint`, `add_variable`, ... API hence it uses `default_copy_to`
* either use `allocate_...` `load...` hence it used `allocate_load`
* or simply implement a custom `MOI.copy_to`.

This is not true that the solver requires to implement Allocate-Load,

Reply:
You made a point. I understand that this paragraph is unclear. Maybe we could just remove it?

Reply:
Yes, it's tricky to explain and should already be covered in "Implementing copy"

Reply:
I agree :) I will make a reference to "Implementing copy" instead, and rescope this PR.